ledger-core/v2/virtual/statemachine/logger.go

```diff
@@ -0,0 +1,210 @@
+	"strings"
+
+	"github.com/insolar/assured-ledger/ledger-core/v2/conveyor"
+	"github.com/insolar/assured-ledger/ledger-core/v2/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/v2/instrumentation/inslogger"
+	"github.com/insolar/assured-ledger/ledger-core/v2/log"
+	"github.com/insolar/assured-ledger/ledger-core/v2/log/global"
+	"github.com/insolar/assured-ledger/ledger-core/v2/log/logfmt"
+)
+
+type ConveyorLogger struct {
+	smachine.StepLoggerStub
+
+	logger log.Logger
+}
+
+func (c ConveyorLogger) CanLogEvent(eventType smachine.StepLoggerEvent, stepLevel smachine.StepLogLevel) bool {
+	return true
+}
+
+func (c ConveyorLogger) CreateAsyncLogger(ctx context.Context, _ *smachine.StepLoggerData) (context.Context, smachine.StepLogger) {
+	return ctx, c
+}
+
+type LogStepMessage struct {
+	*log.Msg
+
+	Message   string
+	Component string `txt:"sm"`
+	TraceID   string `opt:""`
+
+	MachineName interface{} `fmt:"%T"`
+	MachineID   string
+	SlotStep    string
+	From        string
+	To          string `opt:""`
+
+	Error     string `opt:""`
+	Backtrace string `opt:""`
+
+	ExecutionTime  int64 `opt:""`
+	InactivityTime int64 `opt:""`
+}
+
+func getStepName(step interface{}) string {
+	fullName := runtime.FuncForPC(reflect.ValueOf(step).Pointer()).Name()
+	if lastIndex := strings.LastIndex(fullName, "."); lastIndex >= 0 {
+		fullName = fullName[lastIndex+1:]
+	}
+	if lastIndex := strings.LastIndex(fullName, "-"); lastIndex >= 0 {
+		fullName = fullName[:lastIndex]
+	}
+
+	return fullName
+}
+
+func prepareStepName(sd *smachine.StepDeclaration) {
+	if !sd.IsNameless() {
+		return
+	}
+	sd.Name = getStepName(sd.Transition)
+}
+
+func (c ConveyorLogger) LogEvent(data smachine.StepLoggerData, msg interface{}, fields []logfmt.LogFieldMarshaller) {
+	fmt.Println(msg)
+	c.logger.Errorm(msg, fields...)
+}
+
+func (c ConveyorLogger) LogUpdate(stepLoggerData smachine.StepLoggerData, stepLoggerUpdateData smachine.StepLoggerUpdateData) {
+	special := ""
+
+	switch stepLoggerData.EventType {
+	case smachine.StepLoggerUpdate:
+	case smachine.StepLoggerMigrate:
+		special = "migrate "
+	default:
+		panic("illegal value")
+	}
+
+	prepareStepName(&stepLoggerData.CurrentStep)
+	prepareStepName(&stepLoggerUpdateData.NextStep)
+
+	suffix := ""
+	if stepLoggerData.Flags&smachine.StepLoggerDetached != 0 {
+		suffix = " (detached)"
+	}
+
+	if _, ok := stepLoggerData.Declaration.(*conveyor.PulseSlotMachine); ok {
+		return
+	}
+
+	var (
+		backtrace string
+		err       string
+	)
+	if stepLoggerData.Error != nil {
+		if slotPanicError, ok := stepLoggerData.Error.(smachine.SlotPanicError); ok {
+			backtrace = string(slotPanicError.Stack)
+		}
+		err = stepLoggerData.Error.Error()
+	}
+
+	if err != "" {
+		fmt.Println(err)
+	}
+
+	msg := LogStepMessage{
+		Message: special + stepLoggerUpdateData.UpdateType + suffix,
+
+		MachineName: stepLoggerData.Declaration,
+		MachineID:   fmt.Sprintf("%s[%3d]", stepLoggerData.StepNo.MachineID(), stepLoggerData.CycleNo),
```

Comment:
Please avoid formating in logging. Its better to have separate field CycleNo and StepNo